loadgen/loadgen.cc

```diff
@@ -317,14 +317,24 @@ std::vector<QueryMetadata> GenerateQueries(
       ScheduleDistribution<scenario>(settings.target_qps);
 
   // When sample_concatenate_permutation is turned on, pad to a multiple of the
-  // complete dataset to ensure complete fairness.
-  // FIXME: Only override this for Offline; fix after v2.0
-  if (settings.sample_concatenate_permutation &&
-      scenario == TestScenario::Offline &&
-      samples_per_query % loaded_samples.size() != 0) {
-    size_t pad_size =
+  // complete dataset to ensure fairness.
+  auto enable_equal_issue = settings.sample_concatenate_permutation;
+  if (enable_equal_issue)
+  {
+    if (scenario == TestScenario::Offline &&
+      samples_per_query % loaded_samples.size() != 0)
+    {
+      // In offline mode, we pad samples_per_query
+      size_t pad_size =
         (loaded_samples.size() - samples_per_query % loaded_samples.size());
-    samples_per_query += pad_size;
+      samples_per_query += pad_size;
+    }
+    else if (min_queries % loaded_samples.size() != 0)
+    {
+      // In Server, SingleStream, MultiStream mode, the min_queries should be padded
+      size_t pad_size = (loaded_samples.size() - min_queries % loaded_samples.size());
+      min_queries += pad_size;
```

Comment:
This effectively makes the min_query count = dataset size for Server scenario right? (which is reasonable for LLMs) And does equal_issue_mode ensures that all the samples in the loaded_samples are issued exactly once?

Reply:
@arjunsuresh Equal issue mode is to use samples from entire set to be used uniformly, so it can be 1x of sample set size, or 2x or n-times of sample set size. In those cases, each samples would be used once, or twice or n times, but the overall sequence will be randomly shuffled.

Reply:
Thank you @nv-jinhosuh. Yes, I wanted to know if equal issue mode issues all samples in the dataset before repeating.

Reply:
@arjunsuresh if you look at this code: https://github.com/mlcommons/inference/blob/104d85531af7ae2a07ded36dd6e8c7f2c7746d06/loadgen/loadgen.cc#L234
The entire indices (from 0 to loaded_samples.size() - 1) will be shuffled, and the sampling function is just iterating across the shuffled indices. Each query is covered exactly once here.

Reply:
Thank you @nvzhihanj for sharing the code reference - that's exactly what I was looking for. With this code change now GPTJ singlestream will run a minimum of the dataset size number of queries right?

Reply:
That is correct. It will be roughly equal amount of time as running accuracy mode.